modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,144 @@
+package org.tessellation.dag.l1.infrastructure.block
+
+import cats.data.NonEmptyList
+import cats.effect.Sync
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+import cats.syntax.traverse._
+
+import org.tessellation.dag.l1.domain.block.BlockService
+import org.tessellation.dag.l1.domain.block.types.{PulledTips, TipsConfig}
+import org.tessellation.dag.l1.infrastructure.{Success, UnexpectedStateShift, UpdateResult}
+import org.tessellation.dag.types.{BlockReference, DAGBlock}
+import org.tessellation.ext.collection.MapRefUtils.MapRefOps
+import org.tessellation.security.Hashed
+import org.tessellation.security.hash.ProofsHash
+import org.tessellation.security.signature.Signed
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.PosInt
+import io.chrisdavenport.mapref.MapRef
+import io.estatico.newtype.ops._
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object BlockService {
+
+  def make[F[_]: Sync](tipsConfig: TipsConfig): F[BlockService[F]] =
+    MapRef.ofConcurrentHashMap[F, ProofsHash, StoredBlock]().map(make(_, tipsConfig))
+
+  def make[F[_]: Sync](blocks: MapRef[F, ProofsHash, Option[StoredBlock]], tipsConfig: TipsConfig): BlockService[F] =
+    new BlockService[F] {
+      implicit val logger = Slf4jLogger.getLogger[F]
+
+      def areParentsAccepted(block: DAGBlock): F[Map[BlockReference, Boolean]] =
+        block.parent.traverse { br =>
+          blocks(br.hash).get.map(sb => br -> sb.exists(_.isInstanceOf[AcceptedBlock]))
+        }.map(_.toNem.toSortedMap)
+
+      def accept(hashedBlock: Hashed[DAGBlock]): F[Unit] =
```

Comment:
We could assert here that the state is as expected (if we decide that the state can change from waiting to accepted only).

Reply:
That would be a leak of internal representation to the domain. And we need to base on `DAGBlock` and get a `Hashed` of `DAGBlock` to get the `ProofsHash` right. I meant in this comment that I could check during the update that the block that is being accepted is currently in Waiting (so expected state) and only then update. Not sure though if during redownload we won't simply insert blocks as accepted straight away (without the expected shift from waiting -> accepted).

Reply:
hm, I think that the block state could be also a part of a domain.

Reply:
I guess there is no need right now. The `BlockService` uses it internally. Whether the block is kept in the `Waiting/Accepted` classes in the `BlockService` or in any other way isn't important I guess. If L0 needs it for something then we may extract it. Though I think no interface should rely on getting `Accepted` or `Waiting` block. We should just base on `DAGBlock` wrapped in a `Signed` (for external communication) and `Hashed` for local.